### PR TITLE
[docs] Document the API for custom operations

### PR DESCRIPTION
**Pull request by khalatepradnya**
Towards issue #2002 

image

### CODE REVIEW

File: docs/sphinx/api/default_ops.rst [hunk 1]
```diff
@@ -541,4 +541,73 @@ This is a non-linear transformation, and no template overloads are available.
...
+        int main() {
+            auto counts = cudaq::sample(bell_pair);
+            for (auto &[bits, count] : counts) {
+                printf("%s\n", bits.data());
+            }
+        }
+
+.. note:: 
+
+  Custom operations are currently supported in simulation only.
```
Comment (efratshabtai) [inline docs/sphinx/api/default_ops.rst:613]: It might be worth mentioning the backends that are supported, so it's clear what simulation only means. Maybe we can also add what is the user experience when trying to use a custom operator on non-supported backends (error message?)

Comment (khalatepradnya) [reply docs/sphinx/api/default_ops.rst]: Addressed in d06bc6a7673bd326270377e3a48f19847fc40ce8

File: docs/sphinx/api/default_ops.rst [hunk 2]
```diff
@@ -541,4 +541,74 @@ This is a non-linear transformation, and no template overloads are available.
     .. code-block:: cpp
 
         cudaq::qubit qubit;
-        my(qubit);
\ No newline at end of file
+        my(qubit);
+
+
+User-defined Custom Operations
```
Comment (bmhowe23) [inline docs/sphinx/api/default_ops.rst:547]: nit: I would vote for "defined" to be capitalized.
```suggestion
User-Defined Custom Operations
```

File: docs/sphinx/api/default_ops.rst [hunk 3]
```diff
@@ -541,4 +541,74 @@ This is a non-linear transformation, and no template overloads are available.
...
+==============================
+
+Users can define a custom quantum operation by its unitary matrix. First use 
+the API to register a custom operation, outside of a CUDA-Q kernel. Then the 
+operation can be used within a CUDA-Q kernel like any of the built-in operations
+defined above.
+Custom operations are supported on qubits only (`qudit` with `level = 2`). For 
+multi-qubit operations, the matrix is interpreted with MSB qubit ordering.
```
Comment (bmhowe23) [inline docs/sphinx/api/default_ops.rst:555]: I'm not sure that everyone will interpret "MSB qubit ordering" the same way unless we provide a more explicit example. I think that in addition to showing what you have that also showing a more sophisticated example like a combined "XY" gate on 2 qubits (X on qubit 0 and Y on qubit 1) would be helpful. The result is not symmetrical, so it should help users get all of the ordering correct.

File: docs/sphinx/api/default_ops.rst (hunk 2 above)
Comment (khalatepradnya) [reply docs/sphinx/api/default_ops.rst]: Addressed in cc43fff33919ba6d1aa98985efc8e71863f48e62

File: docs/sphinx/api/default_ops.rst (hunk 3 above)
Comment (khalatepradnya) [reply docs/sphinx/api/default_ops.rst]: I added an example to define CNOT operation to clarify this. See cc43fff33919ba6d1aa98985efc8e71863f48e62

File: docs/sphinx/api/default_ops.rst [hunk 4]
```diff
@@ -541,4 +541,123 @@ This is a non-linear transformation, and no template overloads are available.
...
+
+    .. code-block:: python
+
+        import cudaq
+        import numpy as np
+        
+        cudaq.register_operation("my_cnot", np.array([1, 0, 0, 0, 
+                                                      0, 1, 0, 0, 
+                                                      0, 0, 0, 1, 
+                                                      0, 0, 1, 0]))
```
Comment (bmhowe23) [inline docs/sphinx/api/default_ops.rst:624]: Maybe this should be a topic for a different PR, but why do we have to say that a custom operation like this requires controls (i.e. `my_cnot(control, target)`). It seems like a 4x4 matrix should have exactly 2 targets.

Comment (bmhowe23) [reply docs/sphinx/api/default_ops.rst:624]: In other words, since the 4x4 matrix was fully realized, it no longer needs to have "controls" specified in my opinion. I believe the only reason "controls" are distinctly specified from targets is that the underlying simulator can optimize the gate application by not reading or writing to half of the state vector (for 1 control). But if we are supplying a full 4x4 matrix, then I would imagine the underlying simulator has to read/write the full state vector.

So I would expect the invocation of `my_cnot` to be: `my_cnot(qubits[0:2])`.

Comment (khalatepradnya) [reply docs/sphinx/api/default_ops.rst:624]: `my_cnot(qubits[0:2])` - this should work, I called out control and target since it is a controlled-NOT gate. May be not the best example, but I chose one from standard gate set...

File: docs/sphinx/api/default_ops.rst (hunk 3 above)
Comment (khalatepradnya) [reply docs/sphinx/api/default_ops.rst]: Perhaps we should add multiple examples? Can you help me construct this one to convey this...

Comment (bmhowe23) [reply docs/sphinx/api/default_ops.rst]: Yes, will do. It would be a good reason for me to experiment with it anyway 😄

Comment (bmhowe23) [reply docs/sphinx/api/default_ops.rst]: OK, here are some draft updates: https://github.com/NVIDIA/cuda-quantum/commit/3bae23e30a944784c8b7c62ecc56b6638d113e6f. If you like them, feel free to cherry pick that commit into your branch.

On a separate note, we should probably also put the code snippets into the `docs/sphinx/snippets` folder and reference them with `.. literalinclude` similar to how that is handled here: https://github.com/NVIDIA/cuda-quantum/blob/ad367c1bc562a11255fce11a0cecebe88a455d7b/docs/sphinx/using/quick_start.rst?plain=1#L78

The CI runs items in the snippets directory, so if the syntax ever changes, this will ensure that we keep the corresponding docs up to date.

Comment (khalatepradnya) [reply docs/sphinx/api/default_ops.rst]: Addressed in beecd2aa42f953df3c02b1f0ff2f0414af538343

**Review by bmhowe23 — APPROVED** [review]
This LGTM. Thanks, Pradnya!
